### PR TITLE
feat(planner): Push limit down further

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

cherry pick from  [7127](https://github.com/datafuselabs/databend/pull/7127).

And will fix test in this pr.

Fixes https://github.com/datafuselabs/databend/issues/6865

### CODE REVIEW

**Comment by TCeason** [comment]
@AngleNet  Hi angle if you have time, please help me review the cherry-pick is not lost some code.

**Comment by drmingdrmer** [comment]
`main` branch is force pushed(**again** `:(`, to cleanup big object in history), update your working branch please: [To everybody: main branch is force pushed. #7263](https://github.com/datafuselabs/databend/issues/7263)

**Comment by AngleNet** [comment]
> @AngleNet  Hi angle if you have time, please help me review the cherry-pick does not lose some code. Thx.

@TCeason Thank you for taking this over. I would like to review this. I just reviewed. It seems that the `push limit down over sort` is missing.

**Comment by TCeason** [comment]
> > @AngleNet  Hi angle if you have time, please help me review the cherry-pick does not lose some code. Thx.
> 
> @TCeason Thank you for taking this over. I would like to review this. I just reviewed. It seems that the `push limit down over sort` is missing? https://www.github.com/AngleNet/databend/tree/limit/src%2Fquery%2Fservice%2Fsrc%2Fsql%2Foptimizer%2Frule%2Frewrite%2Frule_push_down_limit_sort.rs


Thx for review. Do some modify in this file but forget to add it :(. That's ok now.

**Comment by TCeason** [comment]
> [build_aarch64_macos failed](https://github.com/datafuselabs/databend/runs/8015752657?check_suite_focus=true#logs)

Caused by:
  failed to fetch submodule `testing/arrow-testing` from https://github.com/apache/arrow-testing

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  SecureTransport error: connection closed via error; class=Net (12)

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    fn apply(&self, s_expr: &SExpr, state: &mut TransformState) -> common_exception::Result<()> {
+        let limit: Limit = s_expr.plan().clone().try_into()?;
+        if limit.limit.is_some() {
+            let child = s_expr.child(0)?;
+            let join: LogicalInnerJoin = child.plan().clone().try_into()?;
+            match join.join_type {
+                JoinType::Left | JoinType::Full => {
+                    println!("in push limit join limit is {:?}", limit);
```
Comment (sundy-li) [inline src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs:87]: some prints should be removed.

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_sort.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+///           \
+///          Sort
+///             \
+///              *
+///
+/// Output: Limit
+///           \
+///          Sort
+///             \
+///            Limit
```
Comment (sundy-li) [inline src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_sort.rs:39]: We can't add `limit` before the sort!

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs (hunk 1 above)
Comment (TCeason) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs:87]: Got, I will drop them all.

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_sort.rs (hunk 2 above)
Comment (sundy-li) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_sort.rs:39]: Limit --> sort --> Get

-->

Limit -> Sort --- > Get {Limit, Sort }

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_sort_scan.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+use crate::sql::optimizer::RuleID;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::LogicalGet;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::RelOp;
+use crate::sql::plans::RelOperator;
+use crate::sql::plans::Sort;
+
+/// Input:  Limit
+///           \
```
Comment (sundy-li) [inline src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_sort_scan.rs:30]: comments are wrong.

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_sort_scan.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+}
+
+impl Rule for RulePushDownSortScan {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, s_expr: &SExpr, state: &mut TransformState) -> Result<()> {
+        println!(
```
Comment (sundy-li) [inline src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_sort_scan.rs:67]: .

Comment (TCeason) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_sort_scan.rs:67]: Already delete the println at local code. I will push it when fix all comment.

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_sort.rs (hunk 2 above)
Comment (TCeason) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_sort.rs]: Modify the comment.

File: src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+/// Output:
+///        Left Outer Join
+///             /    \
+///           Limit   *
+///           /
+///          *
+pub struct RulePushDownLimitOuterJoin {
+    id: RuleID,
+    pattern: SExpr,
+}
```
Comment (leiysky) [inline src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs:41]: Is this a legal rule?

For example,
```sql
create table t(a int);
create table t1(a int);
insert into t values(1);
insert into t1 values(1),(1);

select * from t left join t1 on t.a = t1.a limit 1;
 a | a
---+---
 1 | 1
(1 row)

select * from (select * from t limit 1) as t left join t1 on t.a = t1.a;
 a | a
---+---
 1 | 1
 1 | 1
(2 rows)
```
The result is different after this push down.

Comment (TCeason) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs:41]: If want to push limit with join,  maybe the rule need modify to this ?

```
/// Input:      Limit
///               |
///        Left Outer Join
///             /    \
///            *      *
///
/// Output:
///             limit
///              |
///        Left Outer Join
///             /    \
///           Limit   *
///           /
///          *
```

Comment (leiysky) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs:41]: @TCeason Yes, it's possible to push the limit down to both sides of outer join.

Comment (TCeason) [reply src/query/service/src/sql/optimizer/rule/rewrite/rule_push_down_limit_join.rs:43]: Add test in ther: tests/logictest/suites/base/03_dml/03_0010_select_limit_offset

File: src/query/service/tests/it/sql/optimizer/heuristic/testdata/subquery.test [hunk 6]
```diff
@@ -25,25 +25,25 @@ Project: [number (#0)]
     Filter: [(3 (#3)) OR (t.number (#0) > 1)]
         HashJoin: MARK, build keys: [subquery_0 (#0)], probe keys: [subquery_2 (#2)], join filters: []
             HashJoin: INNER, build keys: [t1.number (#1)], probe keys: [subquery_2 (#2)], join filters: []
-                Scan: default.system.numbers, filters: []
-                Scan: default.system.numbers, filters: []
-            Scan: default.system.numbers, filters: []
+                Scan: default.system.numbers, filters: [], Sort: [none], limit: [none]
```
[the remark this answers is not in the document]
Comment (TCeason) [reply src/query/service/tests/it/sql/optimizer/heuristic/testdata/subquery.test:28]: Added in this file. Please review.

File: src/query/service/tests/it/sql/optimizer/heuristic/subquery.rs [hunk 7]
```diff
@@ -99,6 +99,31 @@ pub async fn test_heuristic_optimizer_subquery() -> Result<()> {
...
+        Suite {
+            comment: "# Push down limit join".to_string(),
+            query: "select t.number from numbers(1) as t, numbers(1) as t1 where t.number = (select count(*) from numbers(1) as t2, numbers(1) as t3 where t.number = t2.number) group by t.number order by t.number desc limit 3;".to_string(),
+            rules: DEFAULT_REWRITE_RULES.clone(),
+        },
+        Suite {
+            comment: "# Limit left join with complex expr".to_string(),
+            query: "select * from (select count(t1.number) as c1 from numbers(1) as t1 group by number) as t3 left join (select count(t.number) as c from numbers(2) as t group by number) as t4 on t3.c1=t4.c order by t3.c1 limit 1".to_string(),
+            rules: DEFAULT_REWRITE_RULES.clone(),
+        },
```
Comment (leiysky) [inline src/query/service/tests/it/sql/optimizer/heuristic/subquery.rs:126]: The unit test will be migrated to logictest in #6643 , would you move them into `tests/logictest/suites/mode/standalone/limit.test`?

Comment (TCeason) [reply src/query/service/tests/it/sql/optimizer/heuristic/subquery.rs:126]: absolutely ok~

File: src/query/service/tests/it/sql/optimizer/heuristic/testdata/subquery.test (hunk 6 above)
[the remark this answers is not in the document]
Comment (TCeason) [reply src/query/service/tests/it/sql/optimizer/heuristic/testdata/subquery.test:28]: Modify to tests/logictest/suites/mode/standalone/explain/limit.test

**Review by leiysky — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

File: tests/suites/0_stateless/04_explain/04_0002_explain_v2_cluster.result [hunk 8]
```diff
@@ -2,17 +2,17 @@
...
 └── Project: [a (#0)]
     └── Filter: [t1.a (#0) > 0]
-        └── Scan: default.default.t1, filters: [t1.a (#0) > 0]
+        └── Scan: default.default.t1, filters: [t1.a (#0) > 0], Sort: [none], limit: [none]
 Exchange(Merge)
 └── Filter: [(t1.a (#0) > 3) OR ((t2.a (#2) > 5) AND (t1.a (#0) > 1))]
     └── HashJoin: INNER, build keys: [t2.a (#2)], probe keys: [t1.a (#0)], join filters: []
         ├── Exchange(Hash): keys: [t1.a (#0)]
-        │   └── Scan: default.default.t1, filters: []
+        │   └── Scan: default.default.t1, filters: [], Sort: [none], limit: [none]
```
Comment (sundy-li) [inline tests/suites/0_stateless/04_explain/04_0002_explain_v2_cluster.result:10]: Maybe we should not print the `none` infos.  @leiysky

**Comment by sundy-li** [comment]
Maybe we should not print the `none` infos in another pr.  @leiysky  do you think so?

**Comment by TCeason** [comment]
> Maybe we should not print the `none` infos in another pr. @leiysky do you think so?

Yes a little strange. If none I think print “” maybe better but I’m not sure.

File: src/query/service/src/sql/executor/physical_plan_builder.rs [hunk 9]
```diff
@@ -92,6 +92,19 @@ impl PhysicalPlanBuilder {
                     })
                     .transpose()?;
 
+                let order_by = scan
```
Comment (sundy-li) [inline src/query/service/src/sql/executor/physical_plan_builder.rs:95]: Should build `Expression::Sort`

**Comment by TCeason** [comment]
@Xuanwo This pr will fix main cargo test err with test storages::fuse::pruning::test_block_pruner

**Pull request merged**